waku/v2/protocol/waku_peer_exchange/protocol.nim

```diff
@@ -0,0 +1,209 @@
+      numPeers: numPeers
+    )
+  )
+
+  let res = await wpx.sendPeerExchangeRpcToPeer(rpc, peer)
+  if res.isErr():
+    waku_px_errors.inc(labelValues = [res.error()])
+    return err(res.error())
+
+  return ok()
+
+proc request*(wpx: WakuPeerExchange, numPeers: uint64): Future[WakuPeerExchangeResult[void]] {.async, gcsafe.} =
+  let peerOpt = wpx.peerManager.selectPeer(WakuPeerExchangeCodec)
+  if peerOpt.isNone():
+    waku_px_errors.inc(labelValues = [peerNotFoundFailure])
+    return err(peerNotFoundFailure)
+
+  return await wpx.request(numPeers, peerOpt.get())
+
+proc respond(wpx: WakuPeerExchange, enrs: seq[enr.Record], peer: RemotePeerInfo | PeerId): Future[WakuPeerExchangeResult[void]] {.async, gcsafe.} =
+  var peerInfos: seq[PeerExchangePeerInfo] = @[]
+  for e in enrs:
+    let pi = PeerExchangePeerInfo(
+      enr: e.raw
+    )
+    peerInfos.add(pi)
+
+  let rpc = PeerExchangeRpc(
+    response: PeerExchangeResponse(
+      peerInfos: peerInfos
+    )
+  )
+
+  let res = await wpx.sendPeerExchangeRpcToPeer(rpc, peer)
+  if res.isErr():
+    waku_px_errors.inc(labelValues = [res.error()])
+    return err(res.error())
+
+  return ok()
+
+proc respond*(wpx: WakuPeerExchange, enrs: seq[enr.Record]): Future[WakuPeerExchangeResult[void]] {.async, gcsafe.} =
+  let peerOpt = wpx.peerManager.selectPeer(WakuPeerExchangeCodec)
+  if peerOpt.isNone():
+    waku_px_errors.inc(labelValues = [peerNotFoundFailure])
+    return err(peerNotFoundFailure)
+
+  return await wpx.respond(enrs, peerOpt.get())
+
+proc cleanCache(px: WakuPeerExchange) {.gcsafe.} =
+  px.enrCache.delete(0, CacheCleanWindow-1)
+
+proc runPeerExchangeDiscv5Loop*(px: WakuPeerExchange) {.async, gcsafe.} =
+  ## Runs a discv5 loop adding new peers to the px peer cache
+  if px.wakuDiscv5.isNone():
+    warn "Trying to run discovery v5 (for PX) while it's disabled"
+    return
+
+  info "Starting peer exchange discovery v5 loop"
+
+  while px.wakuDiscv5.get().listening:
+    trace "Running px discv5 discovery loop"
+    let discoveredPeers = await px.wakuDiscv5.get().findRandomPeers()
+    info "Discovered px peers via discv5", count=discoveredPeers.get().len()
+    if discoveredPeers.isOk:
+      for dp in discoveredPeers.get():
+        if dp.enr.isSome() and not px.enrCache.contains(dp.enr.get()):
+          px.enrCache.add(dp.enr.get())
+
+    if px.enrCache.len() >= MaxCacheSize:
+      px.cleanCache()
+
+    ## This loop "competes" with the loop in wakunode2
+    ## For the purpose of collecting px peers, 30 sec intervals should be enough
+    await sleepAsync(30.seconds)
+
+proc getEnrsFromCache(px: WakuPeerExchange, numPeers: uint64): seq[enr.Record] {.gcsafe.} =
+  randomize()
+  if px.enrCache.len() == 0:
+    debug "peer exchange ENR cache is empty"
+    return @[]
+  for i in 0..<min(numPeers, px.enrCache.len().uint64()):
+    let ri = rand(0..<px.enrCache.len())
+    result.add(px.enrCache[ri])
+
+proc init(px: WakuPeerExchange) =
```

Comment:
I would rather rename this procedure: `initProtocolHandler(...)` to avoid mixing it with the exported `init` procedure. It is a more descriptive name.

Reply:
I am already doing it for waku store, see: 
 
https://github.com/status-im/nwaku/blob/refactor-store-protocol/waku/v2/protocol/waku_store/protocol.nim#L201